### PR TITLE
Reconnection logic + resending jsonrpc requests on reconnect

### PR DESCRIPTION
**Pull request by fjoubert**
Opening this PR so that interested parties can have a look so long. I am busy implemented unit tests that I'll commit later, hence the do-not-merge label.

Note: this was branched from user/fjoubert/CB-1790/target-descriptions-from-conf and should be a much smaller PR when that is merged.

### CODE REVIEW

**Comment by fjoubert** [comment]
See https://github.com/ska-sa/katportal/pull/254 for the consumer of the PING heart beat messages.

**Comment by fjoubert** [comment]
This PR is ready for a review and merge - I need to add some more unit tests testing some different disconnect scenarios and will do that in another PR.

File: katportalclient/client.py [hunk 1]
```diff
@@ -217,28 +227,191 @@ def is_connected(self):
         return self._ws is not None
 
     @tornado.gen.coroutine
-    def connect(self):
+    def connect(self, reconnecting=False):
```
Comment (fjoubert) [inline katportalclient/client.py]: Good idea, I'll do that.

File: katportalclient/client.py [hunk 2]
```diff
@@ -217,28 +227,191 @@ def is_connected(self):
...
+                        self._heart_beat_timer.stop()
+                    self._ws = yield websocket_connect(
+                        self.sitemap['websocket'],
+                        on_message_callback=self._websocket_message,
+                        connect_timeout=WS_CONNECT_TIMEOUT)
+                    if reconnecting:
+                        yield self._resend_subscriptions_and_strategies()
+                        self._logger.info("Reconnected :)")
+                    self._heart_beat_timer.start()
+                except:
```
Comment (fjoubert) [inline katportalclient/client.py]: Will do, this is a bad habit I need to unlearn 😅

File: katportalclient/client.py [hunk 3]
```diff
@@ -217,28 +227,191 @@ def is_connected(self):
...
+        that the websocket connection has failed.
+        """
+        self._ws.write_message('PING')
+
     def disconnect(self):
         """Disconnect from the connected websocket server."""
         if self.is_connected:
+            self._disconnect_issued = True
+            self._ws_jsonrpc_cache = []
             self._ws.close()
```
Comment (fjoubert) [inline katportalclient/client.py]: Definitely, good catch.

File: katportalclient/client.py [hunk 4]
```diff
@@ -217,28 +227,191 @@ def is_connected(self):
...
+                self._process_redis_message(msg, msg_id)
+            elif msg_id.startswith('redis-reconnect'):
+                # only resubscribe to namespaces, the server will still
+                # publish sensor value updates to redis because the client
+                # did not disconnect, katportal lost its own connection
+                # to redis
+                yield self._resend_subscriptions()
+            else:
+                self._process_json_rpc_message(msg, msg_id)
+        except:
```
Comment (fjoubert) [inline katportalclient/client.py]: Unforgivable! I'll make it better 😄

File: katportalclient/client.py [hunk 5]
```diff
@@ -217,27 +227,211 @@ def is_connected(self):
...
     def disconnect(self):
         """Disconnect from the connected websocket server."""
         if self.is_connected:
+            self._disconnect_issued = True
+            self._ws_jsonrpc_cache = []
             self._ws.close()
             self._ws = None
             self._logger.debug("Disconnected client websocket.")
+        if self._heart_beat_timer.is_running():
+            self._heart_beat_timer.stop()
```
Comment (ajoubertza) [inline katportalclient/client.py]: Good that you've added this, but it should be before line 295.  The heartbeat time writes to the websocket, so we want to stop it before closing.

Comment (fjoubert) [inline katportalclient/client.py]: Eish, should've thought about that. Changed in 	73de473.

File: katportalclient/client.py [hunk 6]
```diff
@@ -217,28 +227,213 @@ def is_connected(self):
...
+                    self._heart_beat_timer.start()
+                except Exception:
+                    self._logger.exception(
+                        'Could not connect websocket to %s',
+                        self.sitemap['websocket'])
+                    if reconnecting:
+                        self._logger.info(
+                            'Retrying connection in %s seconds...', WS_RECONNECT_INTERVAL)
+                        self._io_loop.call_later(
+                            WS_RECONNECT_INTERVAL, self.connect, True)
```
Comment (fjoubert) [inline katportalclient/client.py]: Yes! Thanks!

File: katportalclient/client.py [hunk 7]
```diff
@@ -217,28 +227,213 @@ def is_connected(self):
...
-                # TODO(TA): check the connect_timeout option
                 self._logger.debug(
                     "Connecting to websocket %s", self.sitemap['websocket'])
-                self._ws = yield websocket_connect(
-                    self.sitemap['websocket'], io_loop=self._io_loop)
-                if self.is_connected:
-                    self._io_loop.add_callback(self._run)
-                else:
+                try:
+                    if self._heart_beat_timer.is_running():
```
Comment (fjoubert) [inline katportalclient/client.py]: Did not have a problem, was just being super safe.

**Comment by ajoubertza** [comment]
Are you still going to add some tests for these changes?

**Comment by fjoubert** [comment]
I am planning on it. Actually forgot a bit about tests after I started working on adding some userlog and auth stuff 🙈

**Comment by fjoubert** [comment]
I merged b-user/tockards/cb-1762/login-logout-functionality into this branch. I will add tests for reconnection logic there. This branch can be merged if we are happy with it.

**Pull request merged**